### PR TITLE
Updates to base class and general specification

### PR DESCRIPTION
**Pull request by valentineap**
This PR implements various changes to the base class. 

- Changing `__init__()` so that default `example_number` is `1` -- it seems unnatural (if Pythonic!) to talk of 'example number 0'.
- Addition of new class methods:
  - `description()` (optional) -- a brief explanation of the active example
  - `model_size()` -- Length of model vector
  - `data_size()` -- Length of data vector
  - `starting_model()` -- A sensible null model, possibly `np.zeros()`
  - `covariance_matrix()` -- A covariance matrix
  - `inverse_covariance_matrix()` (optional) -- Inverse of covariance matrix
  - `misfit(data,pred)` (optional) -- A problem-appropriate measure of data fit
  - `log_likelihood(data,pred)` (optional) -- A problem-appropriate definition of likelihood
  - `log_prior(model)` (optional) -- A problem-appropriate definition of prior
- Renaming of existing class method:
  - `suggested_model()` -> `good_model()` -- I felt this was more intuitive 
- Class methods given @attribute decorator, allowing use as a (read-only) attribute:
  - `description`
  - `model_size`
  - `data_size`
  - `good_model`
  - `starting_model`
  - `data`
  - `covariance_matrix`
  - `inverse_covariance_matrix`
- Removed separate metadata file and instead required certain metadata-like fields to be set in module.

The script `utils/build/validate.py` and the template used by `utils/new_contribution/create_new_contribution.py` have been updated to reflect this.

The `gravity` example has been updated to suit the new structure EXCEPT that `covariance_matrix()` is currently broken. At present the 'real data' is computed at runtime and the noise model depends on the amplitude of the result. Some reconfiguration may be required.

### CODE REVIEW

**Comment by jwhhh** [comment]
Moved metadata attributes into class properties. Instead of
```python
from cofi_espresso import GravityDensity, gravity_density

# when you the methods of the problem
p = GravityDensity()
print(p.good_model())
# when you access the metadata, so you have to import both class and module
print(gravity_density.author_names)
```

We now have everything inside the class
```python
from cofi_espresso import GravityDensity

print(GravityDensity.author_names)
print(GravityDensity().author_names)     # same effect as above
```

The metadata attributes are class attributes, so they are attached to the class itself instead of an instance. They are enforced as abstract as well - i.e., there will be an error immediately when the class `GravityDensity` doesn't have this class level attribute `author_names`.

File: src/cofi_espresso/espresso_problem.py [hunk 1]
```diff
@@ -1,37 +1,238 @@
...
+        
+        d - np.array, shape(data_size) -- A data vector in the same
+            format as output by forward().
+        """
         raise NotImplementedError
-    
+
+    @property
+    @abstractmethod
+    def covariance_matrix(self):
```
Comment (jwhhh) [inline src/cofi_espresso/espresso_problem.py:121]: Is it always necessary? Does it increase the barrier to contributing?

Comment (valentineap) [reply src/cofi_espresso/espresso_problem.py:121]: I guess we could make it optional...

**Comment by valentineap** [comment]
@jwhhh I'll leave this for you to merge if you're happy with it...

**Pull request merged**